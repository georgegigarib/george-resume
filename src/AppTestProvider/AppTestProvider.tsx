import React from 'react'
import { I18nextProvider } from 'react-i18next'
import i18n from '@/locales/i18n.config'
import store from '@/store'
import { Provider } from 'react-redux'
import { BrowserRouter as Router, Routes, Route } from 'react-router-dom'
import { vi } from 'vitest'

vi.mock('swiper/react', () => ({
  Swiper: ({
    children,
    onSlideChange,
  }: {
    children: React.ReactNode
    onSlideChange?: (params: { realIndex: number }) => void
  }) => {
    onSlideChange?.({ realIndex: 0 })
    return <div data-testid="swiper">{children}</div>
  },
  SwiperSlide: ({ children }: { children: React.ReactNode }) => <div data-testid="swiper-slide">{children}</div>,
}))

const AppTestProvider = ({ children }: { children: React.ReactNode }) => {
  return (
    <I18nextProvider i18n={i18n}>
      <Provider store={store}>
        <Router>
          <Routes>
            <Route path="/" element={children} />
          </Routes>
        </Router>
      </Provider>
    </I18nextProvider>
  )
}

export default AppTestProvider
