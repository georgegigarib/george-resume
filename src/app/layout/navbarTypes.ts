export interface boxType {
  startRow: number
  spanRow: number
  startColumn: number
  spanColumn: number
  content: string
}

export enum Sections {
  Dev = '#dev',
  Me = '#me',
  Music = '#music'
}
