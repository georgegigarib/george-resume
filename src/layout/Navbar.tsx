import { useState, useEffect } from "react";
import { useLocation, useNavigate } from "react-router-dom";

const Navbar = () => {
  const [activeHash, setActiveHash] = useState('');
  const [translateClass, setTranslateClass] = useState("");
  const location = useLocation();
  const navigate = useNavigate();

  const scrollToSection = (id: string) => {
    const section = document.querySelector(id) as HTMLElement;
    if (section) {
      const offset =
        section.offsetLeft - window.innerWidth / 2 + section.clientWidth / 2;

      window.scrollTo({
        left: offset,
      });

      const newHash = `${id}`;
      navigate(`${location.pathname}${newHash}`, { replace: true });
    }
  };

  useEffect(() => {
    const hash = location.hash || "#me";
    setActiveHash(hash);

    const getTranslateXForHash = (hash: string) => {
      switch (hash) {
        case "#dev":
          return "translate-x-[48%] xl:translate-x-[52%]";
        case "#me":
          return "-translate-x-0";
        case "#music":
          return "-translate-x-[44%] sm:-translate-x-[40%] md:-translate-x-[45%] xl:-translate-x-[50%]"; 
        default:
          return "translate-x-0";
      }
    };

    setTranslateClass(getTranslateXForHash(hash));
    scrollToSection(hash)
  }, [location.hash]);

  return (
    <nav className="bg-transparent border-transparent  fixed w-full z-20 top-0 start-0 border-b   nav-bottom-oval h-20">
      <div className="max-w-screen-xl flex flex-wrap items-center justify-center mx-auto p-8">
        <div
          className={`justify-between flex mt-2 h-16 xs:gap-28 sm:gap-56 md:gap-80 lg:gap-[450px] xl:gap-[600px] transform ${translateClass}  animated-transform`}
          id="navbar-sticky"
        >
          <a
            onClick={() => scrollToSection("#dev")}
            className={`block rounded md:p-0 ${
              activeHash === "#dev"
                ? "text-white font-bold animated-transform translate-y-[45%] md:translate-y-[45%] scale-130"
                : "text-gray-900 animated-transform rotate-[18deg] translate-y-[10%] sm:rotate-12 md:translate-y-[0%] xl:rotate-3 opacity-40 font-bold s 2xl:translate-y-[20%]"
            } ${activeHash === "#music" ? "-translate-y-[60%] md:-translate-y-[80%] 2xl:-translate-y-[50%]" : ""}`}
            aria-current={activeHash === "#dev" ? "page" : undefined}
          >
            Dev
          </a>
          <a
            onClick={() => scrollToSection("#me")}
            className={`block pt-8 rounded md:p-0 ${
              activeHash === "#me"
                ? "text-white font-bold animated-transform -translate-y-[20%] md:translate-y-[50%] scale-130"
                : "text-gray-900 -translate-y-[40%] sm:-translate-y-[40%] animated-transform md:dark:hover:bg-transparent opacity-40 font-bold"
            } ${
              activeHash === "#music"
                ? "rotate-[18deg]  sm:rotate-6 md:-translate-y-[0%] 2xl:translate-y-[15%] lg:rotate-3"
                : activeHash === "#dev"
                ? "-rotate-[18deg] sm:-rotate-6  md:translate-y-[0%]  2xl:translate-y-[15%] lg:-rotate-3"
                : ""
            }`}
          >
            Me
          </a>
          <a
            onClick={() => scrollToSection("#music")}
            className={`block rounded md:p-0 ${
              activeHash === "#music"
                ? "text-white font-bold animated-transform translate-y-[45%] scale-130 -rotate-1"
                : "text-gray-900 -rotate-[18deg] translate-y-[10%] sm:-rotate-12 md:translate-y-[0%] xl:-rotate-3 animated-transform opacity-40 font-bold 2xl:translate-y-[20%]"
            } ${activeHash === "#dev" ? "-translate-y-[60%] md:-translate-y-[80%] 2xl:-translate-y-[50%]" : ""}`}
          >
            Music
          </a>
        </div>
      </div>
    </nav>
  );
};

export default Navbar;
