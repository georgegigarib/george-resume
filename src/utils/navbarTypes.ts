export interface boxType {
    startRow: number,
    spanRow: number,
    startColumn: number,
    spanColumn: number,
    content: string
}