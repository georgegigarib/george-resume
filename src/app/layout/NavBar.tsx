import { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useLocation } from 'react-router-dom';
import { setActiveHash } from '@/store/hash/activeHashSlice';
import LanguageSwitcher from '@/components/organisms/NavBarDropdownMenu';
import { Sections } from '@/app/layout/navbarTypes';
import { RootState } from "@/store/hash/activeHashSlice";

import DevLink from '@/components/atoms/DevLink';
import MeLink from '@/components/atoms/MeLink.tsx';
import MusicLink from '@/components/atoms/MusicLink';

const Navbar = () => {
  const dispatch = useDispatch();
  const location = useLocation();
  const activeHash = useSelector((state: RootState) => state.hash.activeHash);
  const [translateClass, setTranslateClass] = useState('');
  const [automaticScroll, setAutomaticScroll] = useState(true);

  const moveNavBar = (hash: string): void => {
    const getTranslateXForHash = (hash: string) => {
      switch (hash) {
        case Sections.Dev:
          return 'translate-x-[48%] xl:translate-x-[52%]';
        case Sections.Me:
          return '-translate-x-0';
        case Sections.Music:
          return '-translate-x-[44%] sm:-translate-x-[45%] md:-translate-x-[45%] xl:-translate-x-[50%]';
        default:
          return 'translate-x-0';
      }
    };

    setTranslateClass(getTranslateXForHash(hash));
  };

  const scrollToSection = (id: string, called?: boolean) => {
    const section = document.querySelector(id) as HTMLElement;
    if (section) {
      const offset = section.offsetLeft - window.innerWidth / 2 + section.clientWidth / 2;
      setAutomaticScroll(true);

      window.scrollTo({
        left: offset,
        behavior: 'smooth',
      });

      setTimeout(() => {
        setAutomaticScroll(false);
      }, called ? 0 : 500);
    }
  };

  useEffect(() => {
    const hash = location.hash || Sections.Me;
    dispatch(setActiveHash(hash)); // Despacha la acción para establecer el hash activo
    moveNavBar(hash);

    if (automaticScroll) {
      scrollToSection(hash);
      return;
    }
  }, [location.hash, automaticScroll, dispatch]); // Asegúrate de incluir dispatch como dependencia

  useEffect(() => {
    if (automaticScroll) return;

    const handleScroll = () => {
      const scrollX = window.scrollX;
      const windowWidth = window.innerWidth;
      const documentWidth = document.body.scrollWidth;

      const scrollPercentage = (scrollX / (documentWidth - windowWidth)) * 100;

      if (scrollPercentage > 66) {
        moveNavBar(Sections.Music);
        dispatch(setActiveHash(Sections.Music));
        history.pushState(null, '', Sections.Music);
      } else if (scrollPercentage < 33) {
        moveNavBar(Sections.Dev);
        dispatch(setActiveHash(Sections.Dev));
        history.pushState(null, '', Sections.Dev);
      } else {
        moveNavBar(Sections.Me);
        dispatch(setActiveHash(Sections.Me));
        history.pushState(null, '', Sections.Me);
      }
    };

    window.addEventListener('scroll', handleScroll);
    return () => {
      window.removeEventListener('scroll', handleScroll);
    };
  }, [automaticScroll, dispatch]);

  return (
    <nav className="bg-transparent border-transparent fixed w-full z-20 top-0 start-0 border-b nav-bottom-oval h-12 md:h-20">
      <LanguageSwitcher />

      <div className="max-w-screen-xl flex flex-wrap items-center justify-center mx-auto p-8">
        <div
          className={`justify-between flex mt-0 md:mt-2 h-16 gap-20 xs:gap-28 sm:gap-56 md:gap-80 lg:gap-[450px] xl:gap-[600px] transform ${translateClass} animated-transform`}
          id="navbar-sticky"
        >
          <DevLink scrollToSection={scrollToSection} setAutomaticScroll={setAutomaticScroll} activeHash={activeHash} />
          <MeLink scrollToSection={scrollToSection} setAutomaticScroll={setAutomaticScroll} activeHash={activeHash} />
          <MusicLink scrollToSection={scrollToSection} setAutomaticScroll={setAutomaticScroll} activeHash={activeHash} />
        </div>
      </div>
    </nav>
  );
};

export default Navbar;
