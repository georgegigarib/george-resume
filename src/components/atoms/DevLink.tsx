import React from 'react'
import { Sections } from '@/app/layout/navbarTypes'
import { useTranslation } from 'react-i18next'

const DevLink = ({
  scrollToSection,
  setAutomaticScroll,
  activeHash,
}: {
  scrollToSection: (id: string, called?: boolean) => void
  setAutomaticScroll: React.Dispatch<React.SetStateAction<boolean>>
  activeHash: string
}) => {
  const { t } = useTranslation()

  return (
    <a
      onClick={() => {
        scrollToSection(Sections.Dev, true)
        setAutomaticScroll(false)
      }}
      className={`block rounded md:p-0 ${
        activeHash === Sections.Dev
          ? 'text-white font-bold animated-transform translate-y-[15%] md:translate-y-[45%] scale-130'
          : activeHash === Sections.Music
            ? '-translate-y-[60%] md:-translate-y-[80%] 2xl:-translate-y-[50%] animated-transform'
            : 'text-gray-900 animated-transform rotate-[18deg] -translate-y-[20%] sm:rotate-12 md:translate-y-[0%] xl:rotate-3 opacity-40 font-bold 2xl:translate-y-[20%]'
      }`}
      aria-current={activeHash === Sections.Dev ? 'page' : undefined}
    >
      {t('navbar.dev')}
    </a>
  )
}

export default DevLink
