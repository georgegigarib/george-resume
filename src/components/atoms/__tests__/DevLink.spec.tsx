import { render, fireEvent } from '@testing-library/react';
import { describe, expect, it, vi } from 'vitest';
import DevLink from '@/components/atoms/DevLink';
import { Sections } from '@/app/layout/navbarTypes';
import AppTestProvider from '@/AppTestProvider/AppTestProvider';

describe('DevLink component', () => {
  const mockScrollToSection = vi.fn();
  const mockSetAutomaticScroll = vi.fn();

  const renderComponent = (activeHash: string) =>
    render(
      <AppTestProvider>
        <DevLink scrollToSection={mockScrollToSection} setAutomaticScroll={mockSetAutomaticScroll} activeHash={activeHash} />
      </AppTestProvider>,
    );

  it('matches the snapshot when activeHash is Sections.Dev', () => {
    const { asFragment } = renderComponent(Sections.Dev);
    expect(asFragment()).toMatchSnapshot();
  });

  it('matches the snapshot when activeHash is Sections.Music', () => {
    const { asFragment } = renderComponent(Sections.Music);
    expect(asFragment()).toMatchSnapshot();
  });

  it('matches the snapshot when activeHash is Sections.Me', () => {
    const { asFragment } = renderComponent(Sections.Me);
    expect(asFragment()).toMatchSnapshot();
  });

  it('calls scrollToSection and setAutomaticScroll on click', () => {
    const { getByText } = renderComponent(Sections.Music);
    const linkElement = getByText('dev');

    fireEvent.click(linkElement);

    expect(mockScrollToSection).toHaveBeenCalledWith(Sections.Dev, true);
    expect(mockSetAutomaticScroll).toHaveBeenCalledWith(false);
  });
});
