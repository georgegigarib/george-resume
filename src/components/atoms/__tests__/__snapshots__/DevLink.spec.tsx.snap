// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DevLink component > matches the snapshot when activeHash is Sections.Dev 1`] = `
<DocumentFragment>
  <a
    aria-current="page"
    class="block rounded md:p-0 text-white font-bold animated-transform translate-y-[15%] md:translate-y-[45%] scale-130"
  >
    dev
  </a>
</DocumentFragment>
`;

exports[`DevLink component > matches the snapshot when activeHash is Sections.Me 1`] = `
<DocumentFragment>
  <a
    class="block rounded md:p-0 text-gray-900 animated-transform rotate-[18deg] -translate-y-[20%] sm:rotate-12 md:translate-y-[0%] xl:rotate-3 opacity-40 font-bold 2xl:translate-y-[20%]"
  >
    dev
  </a>
</DocumentFragment>
`;

exports[`DevLink component > matches the snapshot when activeHash is Sections.Music 1`] = `
<DocumentFragment>
  <a
    class="block rounded md:p-0 -translate-y-[60%] md:-translate-y-[80%] 2xl:-translate-y-[50%] animated-transform"
  >
    dev
  </a>
</DocumentFragment>
`;
